### ISSUE TITLE
The TS4023-TS6133 Paradox (noUnusedLocals)

### ISSUE THREAD

**Issue by jeffijoe** [Microsoft/TypeScript#24287]
**TypeScript Version:**  2.8.3

**Search Terms:**  TS6133 TS4023

**Code**

![image](https://user-images.githubusercontent.com/2505474/40305629-926916ec-5cfb-11e8-8410-977b4ba847a2.png)

![image](https://user-images.githubusercontent.com/2505474/40305644-a10d3700-5cfb-11e8-8665-ef1bc9dc36c7.png)

`tsconfig.json`

```json
{
  "compilerOptions": {
    "strict": true,
    "strictNullChecks": true,
    "sourceMap": true,
    "declaration": true,
    "noUnusedLocals": true
  }
}

```

`src/a.ts`

```ts
import { createNamed } from "./b";

export const Value = createNamed();
```

`src/b.ts`

```ts
export interface Named {}

export function createNamed(): Named {
  return {};
}
```

**Expected behavior:**

Preferably that the compiler will know to add the import when emitting declarations (which I am aware is the root cause of the error in the first example).

If that's not possible, then the compiler should not trigger a warning for imported types that are used for declaration emission only.

**Actual behavior:**

The compiler reports an error when not importing types that get implicitly exported. When adding the import to satisfy that requirement, the compiler errors because of the `noUnusedLocals` option. Removing the unused local (the import) causes the first error, hence the reason I aptly labeled this a paradox.

**Playground Link:** This error requires the use of multiple files. Therefore I have attached screenshots and sources that can be used to reproduce the exact setup.

**Related Issues:** #5711, #9944

**Comment by j-oliveras** [comment]
I don't thinks that the error at Value is correct.

Disabling `declaration` solve the error at `Value`. Is this an error with `declaration`?

As a workaround:
```ts
import { createNamed, Named } from "./b";

export const Value: Named = createNamed();
```
(`Named` is used as a explicit type of Value).

**Comment by jeffijoe** [comment]
@j-oliveras I'm aware of the fact this is caused by `declaration` and that I can  add an explicit annotation to work around it. 😄

I'm raising this issue in hopes that we can work out a way to break the paradox, so that as one solves the chain of errors as the compiler suggests, you end up with compilable code.

**Comment by mhegazy** [comment]
The original issue is a duplicate of https://github.com/Microsoft/TypeScript/issues/9944, and should be fixed in 2.9.1.

looks like there is a diffrent bug though, the output should be:

```ts
// a.d.ts
export declare const Value: import("./b").Named;
```

and not 

```ts
// a.d.ts
export declare const Value: Named;

```

**Comment by jeffijoe** [comment]
@mhegazy ah, my bad, I didn't notice any comment indicating that it was going to get fixed.

Apologies for opening a duplicate! 🙏

**Comment by tsofist** [comment]
@mhegazy for **TS@2.9.1** this bug still reproduced with this code (with same `tsconfig`)

b.ts
```ts
export {
    Hash,
    StringHash, StringHash2
};

interface Hash<T> {
    [key: string]: T;
}

type StringHash = Hash<string>; // type NOT OK

interface StringHash2 extends Hash<string> { // interface OK
}
```

a.ts
```ts
import {StringHash, StringHash2} from "./b";

export {
    doSome
}

const MAP: StringHash = {
    a: "a"
};

const MAP2: StringHash2 = {
    a: "a"
};

function doSome(arg1: string,
                arg2 = MAP, //Error:(16, 17) TS4076: Parameter 'arg2' of exported function has or is using name 'Hash' from external module "./b" but cannot be named.
                arg3 = MAP2) { // << TYPE OK
    console.log(arg1, arg2, arg3);
}
```

Should i start a new issue?

**Comment by mhegazy** [comment]
> Should i start a new issue?

yes please

**Comment by tsofist** [comment]
@weswigham Sorry, I did not notice your comment.
Therefore, the defect has already been written. The problem on @next is no longer reproduced (thanks!)
